### PR TITLE
Feature encoding option for custom table AM

### PR DESCRIPTION
**Pull request by gfphoenix78**
### Contributor's Checklist

Here are some reminders and checklists before/when submitting your pull request, please check them:

- [ ] Make sure your Pull Request has a clear title and commit message. You can take [git-commit](https://github.com/cloudberrydb/cloudberrydb/blob/main/.gitmessage) template as a reference.
- [ ] Sign the Contributor License Agreement as prompted for your first-time contribution(*One-time setup*).
- [ ] Learn the [coding contribution guide](https://cloudberrydb.org/contribute/code), including our code conventions, workflow and more.
- [ ] List your communication in the [GitHub Issues](https://github.com/cloudberrydb/cloudberrydb/issues) or [Discussions](https://github.com/orgs/cloudberrydb/discussions) (if has or needed).
- [ ] Document changes.
- [ ] Add tests for the change
- [ ] Pass `make installcheck`
- [ ] Pass `make -C src/test installcheck-cbdb-parallel`
- [ ] Feel free to request `cloudberrydb/dev` team for review and approval when your PR is ready🥳

### CODE REVIEW

File: src/include/access/reloptions.h [hunk 1]
```diff
@@ -300,9 +300,13 @@ extern void validate_and_adjust_options(StdRdOptions *result, relopt_value *opti
...
 /* attribute enconding specific functions */
-extern List *transformColumnEncoding(Relation rel, List *colDefs,
-										List *stenc, List *withOptions,
-										bool rootpartition, bool allowEncodingClause);
+extern void validateAOCOColumnEncodingClauses(List *aocoColumnEncoding);
+extern List *transformColumnEncoding(const TableAmRoutine *tam, Relation rel, List *colDefs,
+										List *stenc, List *withOptions, bool allowEncodingClause,
+										bool createDefaultOne);
+
+List* transfromColumnEncodingAocoRootPartition(List *colDefs, List *stenc, List *withOptions, bool errorOnEncodingClause);
```
Comment (gfphoenix78) [inline src/include/access/reloptions.h:308]: The function handles not only aoco, but also tables, like pax, that support column encodings.

File: src/backend/access/common/reloptions_gp.c [hunk 2]
```diff
@@ -1307,10 +1307,15 @@ encodings_overlap(List *a, List *b)
  * 2. Ensure that each column is referenced either zero times or once.
  * 3. Ensure that the column reference storage clauses do not clash with the
  * 	  gp_default_storage_options
+ * 
+ * Must pass a validate callback.
+ * 1. Caller need validate RELKIND_RELATION that should pass the callback in access method if exist
+ * 2. Caller need validate RELKIND_PARTITIONED_TABLE that will only happend in aoco
  */
 static void
 validateColumnStorageEncodingClauses(List *aocoColumnEncoding,
```
Comment (gfphoenix78) [inline src/backend/access/common/reloptions_gp.c:1316]: rename aocoColumnEncoding, not only for aoco.

File: src/backend/commands/tablecmds.c [hunk 3]
```diff
@@ -1022,29 +1022,36 @@ DefineRelation(CreateStmt *stmt, char relkind, Oid ownerId,
...
+		 * But for other access method which support encoding clause, still can't
+		 * pass the encoding clause from root partition.
 		 */
-		stmt->attr_encodings = transformColumnEncoding(NULL /* Relation */, 
-								schema,
+		stmt->attr_encodings = transfromColumnEncodingAocoRootPartition(schema,
 								stmt->attr_encodings,
 								stmt->options,
-								relkind == RELKIND_PARTITIONED_TABLE,
 								!AMHandlerIsAoCols(amHandlerOid)
```
Comment (gfphoenix78) [inline src/backend/commands/tablecmds.c:1052]: pax with encoding option?

File: src/backend/access/appendonly/appendonlyam_handler.c [hunk 4]
```diff
@@ -2394,6 +2402,8 @@ static const TableAmRoutine ao_row_methods = {
 
 	.amoptions = ao_amoptions,
 	.swap_relation_files = appendonly_swap_relation_files,
+
+	.transform_column_encoding_clauses = appendonly_transform_column_encoding_clauses,
```
Comment (gfphoenix78) [inline src/backend/access/appendonly/appendonlyam_handler.c:2406]: remove callback.

File: src/backend/commands/tablecmds.c [hunk 5]
```diff
@@ -8294,15 +8303,15 @@ ATExecAddColumn(List **wqueue, AlteredTableInfo *tab, Relation rel,
 	 * If there's an encoding clause for non AO/CO tables, we'll throw an error 
 	 * in the function (indicated by errorOnEncodingClause == true).
 	 */
-	enc = transformColumnEncoding(rel, list_make1(colDef), 
+	enc = transformColumnEncoding(tam /* TableAmRoutine */, rel, list_make1(colDef),
```
Comment (gfphoenix78) [inline src/backend/commands/tablecmds.c:8306]: Add comment: aoco will create default encoding clause, pax will not.

File: src/backend/access/common/reloptions_gp.c [hunk 6]
```diff
@@ -1634,7 +1635,15 @@ List* transformColumnEncoding(Relation rel, List *colDefs, List *stenc, List *wi
...
-			deflt->encoding = transformStorageEncodingClause(tmpenc, false);
+
+			Assert(tam);
+			/*
+			 * if current am not inmplement transform_column_encoding_clauses
+			 * then tmpenc not null but no need fill with options.
+			 */
+			if (tam->transform_column_encoding_clauses) {
+				deflt->encoding = tam->transform_column_encoding_clauses(rel, tmpenc, false, false);
+			}
```
Comment (avamingli) [inline src/backend/access/common/reloptions_gp.c:1646]: pg style:
if 
{
}

File: src/backend/access/common/reloptions_gp.c [hunk 7]
```diff
@@ -1657,6 +1666,111 @@ List* transformColumnEncoding(Relation rel, List *colDefs, List *stenc, List *wi
...
+				if (deflt)
+					encoding = copyObject(deflt->encoding);
+				else
+				{
+					if (d->typeName) {
+						/* get encoding by type, still need do transform and validate */
+						encoding = get_type_encoding(d->typeName);
+						if (tam && tam->transform_column_encoding_clauses)
+							encoding = tam->transform_column_encoding_clauses(rel, encoding, true, true);
+					}
```
Comment (avamingli) [inline src/backend/access/common/reloptions_gp.c:1695]: Do we need to check transform_column_encoding_clauses and transform_column_encoding_clauses is NULL here?

**Review by avamingli — APPROVED** [review]

File: src/backend/access/common/reloptions_gp.c [hunk 8]
```diff
@@ -1580,14 +1568,27 @@ find_crsd(const char *column, List *stenc)
...
+		return result;
+
+	if (stenc) {
+		if (errorOnEncodingClause)
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					errmsg("ENCODING clause only supported with column oriented tables")));
+		/*
+		 * if no `tam` or no encoding callback in am
+		 * then `errorOnEncodingClause` will be true
```
[the remark this answers is not in the document]
Comment (gfphoenix78) [reply src/backend/access/common/reloptions_gp.c:1587]: no, if no tam or no encoding callback in tam, then the tam doesn't support encoding clause, so the function should raise an error.

File: src/backend/access/common/reloptions_gp.c (hunk 7 above)
Comment (gfphoenix78) [reply src/backend/access/common/reloptions_gp.c:1695]: It's hard to tell.

**Review by avamingli — APPROVED** [review]
LGTM.

**Pull request merged**